### PR TITLE
feat(docker): add support for `amazonlinux`

### LINKED ISSUE #3401: Test `amazonlinux` in the `docker` workflow
We should test:
- Amazon Linux 2
- Amazon Linux 2022

on `amd64` and `arm64`.

### PR DESCRIPTION
**Pull request by jkomyno**
Close https://github.com/prisma/ecosystem-tests/issues/3401.

New OS support (Amazon Linux 2 and Amazon Linux 2022 on `amd64` and `arm64`):
- [x] `amazonlinux-2-amd64-openssl-1.0.x`
- [x] `amazonlinux-2-arm64-openssl-1.0.x`
- [x] `amazonlinux-2022-amd64-openssl-3.0.x`
- [x] `amazonlinux-2022-arm64-openssl-3.0.x`

See [relevant changed files](https://github.com/prisma/ecosystem-tests/pull/3402/files?file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.sh&file-filters%5B%5D=.yaml&file-filters%5B%5D=No+extension&show-viewed-files=true).

### CODE REVIEW

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+RUN yum update -y \
+  && yum install -y tar xz
+
+# Install Node.js
+RUN curl -fsSLO --compressed "https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-${ARCH}.tar.xz" \
+  && tar -xJf "node-v${NODE_VERSION}-linux-${ARCH}.tar.xz" -C /usr/local --strip-components=1 --no-same-owner \
+  && rm "node-v${NODE_VERSION}-linux-${ARCH}.tar.xz" \
+  && npm i -g yarn
+
+# Both OpenSSL 1.0 and OpenSSL 3.0 are expected to be on this system
```
Comment (janpio) [inline docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: Interesting, was not aware of that.
Would that mean a Engine with a OpenSSL dynamic linking would also work there?

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# syntax=docker/dockerfile:1
+
+# substitute for centos
+FROM amazonlinux:latest
+
+# Node.js 18 isn't compatible with Amazon Linux due to glibc version. It would throw with:
+#   node: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by node)
+#   node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by node)
```
Comment (Jolg42) [inline docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: Oh I didn't know about this! 
So Node.js v18 requires GLIBC_2.27 at minimum but Amazon Linux 2 comes with 2.26

Comment (janpio) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: Oh wow. So no node18 on Lambda? That can not be possible, right?

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile (hunk 1 above)
Comment (jkomyno) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: Thanks for raising that to my attention. First of all, `libssl3.so` is misleading. I inadvertently treated it as `openssl-3.0.x`, but it's not that. Only `openssl-1.0.x` is installed in `amazonlinux` by default.

> Would that mean a Engine with a OpenSSL dynamic linking would also work there?

Apparently, yes. E.g., if we install `libssl-3.0.x` and manually specify we should use an engine like `debian-openssl-3.0.x`, the system libraries shipped with `libssl.so.3` would be found. Currently, it's unclear to me whether that'd work with `libssl-1.1.x` as well.

I've tried locally using `amazonlinux` with a `sqlite` schema using `binaryTargets = ["native", "debian-openssl-1.0.x", "debian-openssl-3.0.x"]`, where `native` is resolved as `rhel-openssl-1.0.x`.

Using `ldd`, we can see the dynamically linked libraries for each query engine. We can see that the `*-openssl-3.0.x` engines try to dynamically load `libssl.so.3` and `libcrypto.so.3`, which isn't installed:

- `ldd libquery_engine-debian-openssl-3.0.x.so.node`

  ```sh
  libz.so.1 => /lib64/libz.so.1 (0x0000004002a87000)
  libssl.so.3 => not found
  libcrypto.so.3 => not found
  libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x0000004002c9e000)
  librt.so.1 => /lib64/librt.so.1 (0x0000004002eb4000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x00000040030bc000)
  libm.so.6 => /lib64/libm.so.6 (0x00000040032da000)
  libdl.so.2 => /lib64/libdl.so.2 (0x000000400361c000)
  libc.so.6 => /lib64/libc.so.6 (0x0000004003820000)
  /lib64/ld-linux-x86-64.so.2 (0x0000004000000000)
  ```

- `ldd libquery_engine-debian-openssl-1.0.x.so.node`

  ```sh
  libz.so.1 => /lib64/libz.so.1 (0x0000004002cea000)
  libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x0000004002f01000)
  librt.so.1 => /lib64/librt.so.1 (0x0000004003117000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x000000400331f000)
  libm.so.6 => /lib64/libm.so.6 (0x000000400353d000)
  libdl.so.2 => /lib64/libdl.so.2 (0x000000400387d000)
  libc.so.6 => /lib64/libc.so.6 (0x0000004003a81000)
  /lib64/ld-linux-x86-64.so.2 (0x0000004000000000)
  ```

- `ldd libquery_engine-rhel-openssl-3.0.x.so.node`

  ```sh
  libz.so.1 => /lib64/libz.so.1 (0x00000040029c7000)
  libssl.so.3 => not found
  libcrypto.so.3 => not found
  libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x0000004002bde000)
  librt.so.1 => /lib64/librt.so.1 (0x0000004002df4000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x0000004002ffc000)
  libm.so.6 => /lib64/libm.so.6 (0x000000400321a000)
  libdl.so.2 => /lib64/libdl.so.2 (0x000000400355c000)
  libc.so.6 => /lib64/libc.so.6 (0x0000004003760000)
  /lib64/ld-linux-x86-64.so.2 (0x0000004000000000)
  ```

- `ldd libquery_engine-rhel-openssl-1.0.x.so.node`

  ```sh
  libz.so.1 => /lib64/libz.so.1 (0x0000004002c27000)
  libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x0000004002e3e000)
  librt.so.1 => /lib64/librt.so.1 (0x0000004003054000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x000000400325c000)
  libm.so.6 => /lib64/libm.so.6 (0x000000400347a000)
  libdl.so.2 => /lib64/libdl.so.2 (0x00000040037ba000)
  libc.so.6 => /lib64/libc.so.6 (0x00000040039be000)
  /lib64/ld-linux-x86-64.so.2 (0x0000004000000000)
  ```

If I install `libssl-3.0.x` manually, and run `ldd` again on the `*-openssl-3.0.x` engines, I see that the dynamic link is satisfied:

- `ldd libquery_engine-debian-openssl-3.0.x.so.node`

  ```sh
  libz.so.1 => /lib64/libz.so.1 (0x0000004002a89000)
  libssl.so.3 => /lib64/libssl.so.3 (0x0000004002ca0000)
  libcrypto.so.3 => /lib64/libcrypto.so.3 (0x0000004002f42000)
  libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00000040035a9000)
  librt.so.1 => /lib64/librt.so.1 (0x00000040037bf000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x00000040039c7000)
  libm.so.6 => /lib64/libm.so.6 (0x0000004003be5000)
  libdl.so.2 => /lib64/libdl.so.2 (0x0000004003f27000)
  libc.so.6 => /lib64/libc.so.6 (0x000000400412b000)
  /lib64/ld-linux-x86-64.so.2 (0x0000004000000000)
  ```

- `ldd libquery_engine-rhel-openssl-3.0.x.so.node`

  ```sh
  libz.so.1 => /lib64/libz.so.1 (0x00000040029c9000)
  libssl.so.3 => /lib64/libssl.so.3 (0x0000004002be0000)
  libcrypto.so.3 => /lib64/libcrypto.so.3 (0x0000004002e82000)
  libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00000040034e9000)
  librt.so.1 => /lib64/librt.so.1 (0x00000040036ff000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x0000004003907000)
  libm.so.6 => /lib64/libm.so.6 (0x0000004003b25000)
  libdl.so.2 => /lib64/libdl.so.2 (0x0000004003e67000)
  libc.so.6 => /lib64/libc.so.6 (0x000000400406b000)
  /lib64/ld-linux-x86-64.so.2 (0x0000004000000000)
  ```

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile (hunk 2 above)
Comment (jkomyno) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: So, after reading [this](https://aws.amazon.com/about-aws/whats-new/2022/11/aws-lambda-support-node-js-18/) (AWS Lambda adds support for Node.js 18), I've figured that Amazon's Docker image naming is more inconsistent than anticipated.

- The Docker image for Amazon Linux 2 is `amazonlinux` (`:latest`, `:2`)
- The Docker image for Amazon Linux 2022 is `public.ecr.aws/amazonlinux/amazonlinux:2022`

From the [FAQs](https://aws.amazon.com/linux/amazon-linux-2022/faqs/):

> Amazon Linux 2022 (AL2022) is a general-purpose rpm-based Linux distribution and a successor to Amazon Linux 2. AL2022 simplifies planning for operating system upgrades. Starting with AL2022, a new major version of Amazon Linux is released every two years, includes minor quarterly releases, and comes with five years of long-term support. Amazon Linux 2022 integrates with AWS services and is designed to be deployed at scale in the cloud.

Comment (jkomyno) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: I will add both images to this PR.

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile (hunk 1 above)
Comment (janpio) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: > Apparently, yes. E.g., if we install libssl-3.0.x and manually specify we should use an engine like debian-openssl-3.0.x, the system libraries shipped with libssl.so.3 would be found.

Ok, when we manually install OpenSSL 3 then that makes sense.

I was still thinking about the unmodified Amazing Linux in Lambda (where you definitely do not want to install stuff manually).

So generally the answer to my question I think still is: Nope, that was a mistake in this file which now got fixed.

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile (hunk 2 above)
Comment (janpio) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: Wow. Which one runs on Lambda by default?

**Comment by Jolg42** [comment]
so here it's testing https://aws.amazon.com/amazon-linux-2/ 

@jkomyno It looks like we could add a test for Amazon Linux 2022 (maybe in separate PR) but it's currently in Preview

Amazon Linux 2022
https://docs.aws.amazon.com/linux/al2022/ug/core-glibc.html
https://docs.aws.amazon.com/linux/al2022/ug/compare-al2-to-AL2022.html
Screenshot 2023-01-23 at 16 58 43
> The default software package management tool on Amazon Linux 2022 is DNF. DNF is the successor to YUM, the package management tool in Amazon Linux 2.

> Amazon Linux 2022 is based on Fedora, but isn't directly compatible with any particular release of Fedora. Amazon Linux 2 was built on several upstream sources including Fedora.

> Amazon Linux 2022 features the Open Secure Sockets Layer version 3 (OpenSSL 3) cryptography toolkit. Amazon Linux 2022 uses the Secure Sockets Layer (SSL v2/v3) and Transport Layer Security (TLS v1) network protocols. It also uses the required cryptography standards.
> By default, Amazon Linux 2 comes with OpenSSL 1.0.2. You can build applications against OpenSSL 1.1.1.
 
New cadence
https://docs.aws.amazon.com/linux/al2022/ug/release-cadence.html
Screenshot 2023-01-23 at 16 59 35

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile (hunk 2 above)
Comment (jkomyno) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: Note: this is the same information posted by @Jolg42 in [this comment](https://github.com/prisma/ecosystem-tests/pull/3402#issuecomment-1400686891) a few minutes later.

**Comment by jkomyno** [comment]
`alpine:3.17 amd64` is failing due to
> Error: P1001: Can't reach database server at `e2e-tests-postgres.cdoyhcosd7km.us-east-1.rds.amazonaws.com`:`5432`

File: docker/amazonlinux-2-arm64-openssl-1.0.x/README.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# OS support: amazonlinux-2-arm64-openssl-1.0.x
+
+- Base Docker image: `archlinux:2`
```
Comment (Jolg42) [inline docker/amazonlinux-2-arm64-openssl-1.0.x/README.md]:
```suggestion
- Base Docker image: `amazonlinux:2`
```

Comment (jkomyno) [reply docker/amazonlinux-2-arm64-openssl-1.0.x/README.md]: Oops

**Review by Jolg42 — APPROVED** [review]
Nice one! 🙌🏼

File: docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile (hunk 1 above)
Comment (jkomyno) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: On Amazon Linux 2022 `openssl-3.0.x` is installed by default, so dynamically linking for `libssl` already applies

Comment (jkomyno) [reply docker/amazonlinux-latest-amd64-openssl-1.0.x/Dockerfile]: (I will mark this conversation as resolved to merge the PR. If anything else seems missing to you, feel free to add a new comment after the PR is merged).

**Comment by Jolg42** [comment]
Added a note on [Slack](https://prisma-company.slack.com/archives/C04DG9JL6TV/p1674495528294259)

Maybe we should actually test the following if we want to test something close to AWS Lambda:
https://github.com/aws/aws-lambda-base-images/tree/nodejs18.x

Because AWS Lambda even for Node.js 18 is running on Amazon Linux 2
https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html

**Pull request merged**